plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,332 @@ extern const char* const state_history_plugin_abi;
+               }
+
+               if(status_requests.size())
+                  current_status_result = fill_current_status_result();
+               co_return;
+            }, boost::asio::use_awaitable);
+
+            //if there is nothing to send, go to sleep
+            if(status_requests.empty() && !block_to_send) {
+               co_await wake_timer.async_wait();
+               continue;
+            }
 
-      // on exception allow session to be destroyed
+            //send replies to all send status requests first
+            for(const bool status_request_is_v1 : status_requests) {
+               if(status_request_is_v1 == false) //v0 status request, gets a v0 status result
+                  co_await stream.async_write(boost::asio::buffer(fc::raw::pack(state_result((get_status_result_v0)current_status_result))));
+               else
+                  co_await stream.async_write(boost::asio::buffer(fc::raw::pack(state_result(current_status_result))));
+            }
 
-      fc_ilog(plugin.get_logger(), "Closing connection from ${a}", ("a", description));
-      session_mgr.remove( this->shared_from_this(), active_entry );
+            //and then send the block
+            if(block_to_send) {
+               const fc::unsigned_int get_blocks_result_variant_index = block_to_send->is_v1_request ?
+                                                                        state_result(get_blocks_result_v1()).index() :
+                                                                        state_result(get_blocks_result_v0()).index();
+               co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(get_blocks_result_variant_index)));
+               co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(block_to_send->blocks_result_base)));
+
+               //accessing the _logs here violates the rule that those should only be accessed on the main thread. However, we're
+               // only calling get_unpacked_entry() on it which assumes the mutex is held by the locked_decompress_stream. So this is
+               // "safe" in some aspects but can deadlock
+               co_await write_log_entry(block_to_send->trace_stream, trace_log, block_to_send->this_block_num);
+               co_await write_log_entry(block_to_send->state_stream, chain_state_log, block_to_send->this_block_num);
+               if(block_to_send->is_v1_request)
+                  co_await write_log_entry(block_to_send->finality_stream, finality_data_log, block_to_send->this_block_num);
+
+               co_await stream.async_write_some(true, boost::asio::const_buffer());
+            }
+         }
+      });
    }
+
+private:
+   ///these items must only ever be touched by the session's strand
+   Executor                          strand;
+   coro_throwing_stream              stream;
+   coro_nonthrowing_steadytimer      wake_timer;
+   unsigned                          coros_running = 0;
```

Comment:
I know the manual counting of how many coroutines are running, to then manually fire off a callback that the `session` can be destroyed, looks suspect as not being very C++-ish. I tried some approaches with `shared_ptr` to do automatic cleanup but all attempts resulted in very spooky crashes and/or sanitizer errors.

The manual counting is potentially error prone, but it allows crafting the rules of when the `session` is destroyed (critically, when no coroutine's context is "alive" at all) very deliberately.